### PR TITLE
Adapt adapter pattern for cart

### PR DESCRIPTION
**Pull request by dunqan**
Use Connectors, Adapters and Normalizers for product search backend calls

Closes GH-2035

### CODE REVIEW

File: projects/core/src/cart/occ/occ-cart.adapter.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import { ConverterService } from '../../util/converter.service';
+import { CART_NORMALIZER } from '../connectors/cart/converters';
+
+// for mini cart
+const BASIC_PARAMS =
+  'DEFAULT,deliveryItemsQuantity,totalPrice(formattedValue),' +
+  'entries(totalPrice(formattedValue),product(images(FULL)))';
+
+// for cart details page
+const DETAILS_PARAMS =
```
Comment (tobi-or-not-tobi) [inline projects/core/src/cart/occ/occ-cart.adapter.ts]: while we're here, can we move the endpoint configuration to the new config?

File: projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { Observable } from 'rxjs';
+import { Address, DeliveryMode } from '../../../occ/occ-models/occ.models';
+
+export abstract class CartDeliveryAdapter {
```
Comment (tobi-or-not-tobi) [inline projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts]: Should the address related methods not be part of an address service? Or is this specific for the cart API?

File: projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+   * Abstract method used to set delivery mode on cart
+   *
+   * @param userId
+   * @param cartId
+   * @param deliveryModeId
+   */
+  abstract setMode(
+    userId: string,
+    cartId: string,
+    deliveryModeId: string
```
Comment (tobi-or-not-tobi) [inline projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts]: Can we make the deliveryMode tye safe? (optionally also allow for string)?

File: projects/core/src/cart/connectors/payment/cart-payment.connector.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import { PaymentDetails } from '../../../occ/occ-models/occ.models';
+import { CartPaymentAdapter } from './cart-payment.adapter';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CartPaymentConnector {
+  constructor(private adapter: CartPaymentAdapter) {}
+
+  public createDetails(
```
Comment (tobi-or-not-tobi) [inline projects/core/src/cart/connectors/payment/cart-payment.connector.ts]: just `create()` ?

File: projects/core/src/cart/occ/occ-cart-payment.adapter.ts [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+      ' ' +
+      paymentDetails.billingAddress.line2;
+    params[mappingLabels['hybris_billTo_city']] =
+      paymentDetails.billingAddress.town;
+    params[mappingLabels['hybris_billTo_postalcode']] =
+      paymentDetails.billingAddress.postalCode;
+    return params;
+  }
+
+  private extractPaymentDetailsFromHtml(html: string): any {
```
Comment (tobi-or-not-tobi) [inline projects/core/src/cart/occ/occ-cart-payment.adapter.ts]: this makes me cry... but i'm sure it's not coming from this PR.

File: projects/core/src/cart/occ/occ-cart.adapter.ts (hunk 1 above)
Comment (dunqan) [reply projects/core/src/cart/occ/occ-cart.adapter.ts]: There is a separate ticket waiting for doing this part AFAIK

File: projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts (hunk 2 above)
Comment (dunqan) [reply projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts]: It looks like it is specific for a cart, to be more precise: `users/user_id/carts/cart_id/addresses/delivery`

File: projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts (hunk 3 above)
Comment (dunqan) [reply projects/core/src/cart/connectors/delivery/cart-delivery.adapter.ts]: Hmm... not sure. We get delivery modes from backend, so it looks that they are backend driven and code could be any string.

File: projects/core/src/cart/occ/occ-cart-payment.adapter.ts (hunk 5 above)
Comment (dunqan) [reply projects/core/src/cart/occ/occ-cart-payment.adapter.ts]: Yes, you are right. Fortunately, it is not a part of public API, but only implementation details of OccCartPaymentAdapter

File: projects/core/src/cart/connectors/payment/cart-payment.connector.ts (hunk 4 above)
Comment (dunqan) [reply projects/core/src/cart/connectors/payment/cart-payment.connector.ts]: Changed `createDetails` to `create` and `setDetails` to `set`,

**Pull request merged**